### ISSUE TITLE
can create a project/namespace with a hyphen but builds blow up on the hyphen

### ISSUE THREAD

**Issue by thoraxe** [openshift/origin#933]
Given the following project:

```json
{
  "kind": "Project",
  "apiVersion": "v1beta1",
  "metadata": {
    "name": "sinatra-project",
    "labels": {
      "name": "sinatra-project"
    },
    "annotations": {
      "description": "This is a simple project to demonstrate Ruby+STI on OpenShift v3"
    }
  },    
  "displayName": "Hello Sinatra",
}
```

When you try to do a build related to this project, the build fails with the following error:

```
[root@ose3-master simple-openshift-sinatra-sti]# osc --namespace=sinatra-project build-logs a1aa7e35-ad82-11e4-8f5f-525400b33d1d
2015-02-05T22:04:10.285998866Z Cloning into '/tmp/sti974342869/upload/src'...
2015-02-05T22:04:11.829370924Z Already on 'master'
2015-02-05T22:04:36.850229019Z I0205 17:04:36.814960       1 build.go:160] Successfully built 172.30.17.158:5001/sinatra-project/simple-openshift-sinatra-sti
2015-02-05T22:04:39.862687558Z 2015/02/05 17:04:39 Build error: API error (500): Invalid namespace name (sinatra-project), only [a-z0-9_] are allowed, size between 4 and 30
```

I am using the app here:
https://github.com/thoraxe/simple-openshift-sinatra-sti

And using `openshift ex generate` inside its folder after removing the Dockerfile. In other words:

```
rm -f Dockerfile
openshift ex generate | osc apply --namespace=sinatra-test -f -
```

I don't understand how the namespace can be invalid if you can not only create it but assign stuff to it.

**Comment by derekwaynecarr** [comment]
This is a bug in project validation that got out of synch with upstream.

A project name must be a DNS Subdomain, will create a fix.

**Comment by derekwaynecarr** [comment]
Actually, this may not be related to the project validation (which was still not right), but with the build validation.

**Comment by smarterclayton** [comment]
This is the docker repository validation.  Docker doesn't let you push to dashed namespaces or names right now, and we're working with upstream to get the v2 registry to support them.

**Comment by thoraxe** [comment]
Understood, but it still is kinda a problem so either:

1) We need to temporarily invalidate projects/namespaces with hyphens or 

2) We need openshift to employ some logic such that it doesn't end up trying to build an image with a hyphen just because the project has a hyphen

Or something??

**Comment by liggitt** [comment]
@ncdc is this fixed with the v2 registry?

**Comment by ncdc** [comment]
@liggitt yes, with Docker 1.6 and the v2 registry, it's possible to have a project with a hyphen in the name.

```
vagrant@localhost:~/go/src/github.com/openshift/origin (master) osc build-logs ruby-sample-build-1
2015-05-08T13:42:40.135124964Z E0508 13:42:40.078640       1 cfg.go:50] /root/.dockercfg: stat /root/.dockercfg: no such file or directory
2015-05-08T13:42:41.569642922Z I0508 13:42:41.569346       1 sti.go:371] ---> Installing application source
2015-05-08T13:42:41.576069381Z I0508 13:42:41.576014       1 sti.go:371] ---> Building your Ruby application from source
2015-05-08T13:42:41.576245925Z I0508 13:42:41.576048       1 sti.go:371] ---> Running 'bundle install --deployment'
2015-05-08T13:42:45.559243716Z I0508 13:42:45.559149       1 sti.go:371] Fetching gem metadata from https://rubygems.org/..........
2015-05-08T13:42:47.283688061Z I0508 13:42:47.283422       1 sti.go:371] Installing rake (10.3.2)
2015-05-08T13:42:47.494412942Z I0508 13:42:47.494146       1 sti.go:371] Installing i18n (0.6.11)
2015-05-08T13:42:48.684919535Z I0508 13:42:48.684622       1 sti.go:371] Installing json (1.8.1)
2015-05-08T13:42:48.885666401Z I0508 13:42:48.885509       1 sti.go:371] Installing minitest (5.4.2)
2015-05-08T13:42:49.146736133Z I0508 13:42:49.146474       1 sti.go:371] Installing thread_safe (0.3.4)
2015-05-08T13:42:49.422793758Z I0508 13:42:49.422523       1 sti.go:371] Installing tzinfo (1.2.2)
2015-05-08T13:42:49.915923096Z I0508 13:42:49.903964       1 sti.go:371] Installing activesupport (4.1.7)
2015-05-08T13:42:50.044802650Z I0508 13:42:50.044419       1 sti.go:371] Installing builder (3.2.2)
2015-05-08T13:42:50.216817402Z I0508 13:42:50.216595       1 sti.go:371] Installing activemodel (4.1.7)
2015-05-08T13:42:50.415039899Z I0508 13:42:50.414773       1 sti.go:371] Installing arel (5.0.1.20140414130214)
2015-05-08T13:42:50.855562365Z I0508 13:42:50.855315       1 sti.go:371] Installing activerecord (4.1.7)
2015-05-08T13:42:52.738356748Z I0508 13:42:52.738073       1 sti.go:371] Installing mysql2 (0.3.16)
2015-05-08T13:42:53.064849760Z I0508 13:42:53.064631       1 sti.go:371] Installing rack (1.5.2)
2015-05-08T13:42:53.205495713Z I0508 13:42:53.205256       1 sti.go:371] Installing rack-protection (1.5.3)
2015-05-08T13:42:53.378772123Z I0508 13:42:53.378550       1 sti.go:371] Installing tilt (1.4.1)
2015-05-08T13:42:53.764105880Z I0508 13:42:53.763872       1 sti.go:371] Installing sinatra (1.4.5)
2015-05-08T13:42:53.893412118Z I0508 13:42:53.893325       1 sti.go:371] Installing sinatra-activerecord (2.0.3)
2015-05-08T13:42:53.893583860Z I0508 13:42:53.893352       1 sti.go:371] Using bundler (1.3.5)
2015-05-08T13:42:53.894704526Z I0508 13:42:53.894433       1 sti.go:371] Your bundle is complete!
2015-05-08T13:42:53.894717135Z I0508 13:42:53.894453       1 sti.go:371] It was installed into ./bundle
2015-05-08T13:42:53.911447970Z I0508 13:42:53.911209       1 sti.go:371] ---> Cleaning up unused ruby gems
2015-05-08T13:42:58.072319224Z E0508 13:42:58.072032       1 cfg.go:50] /root/.dockercfg: stat /root/.dockercfg: no such file or directory
2015-05-08T13:42:58.072344719Z I0508 13:42:58.072085       1 sti.go:80] Pushing 172.30.7.46:5000/andy-test/origin-ruby-sample image ...
2015-05-08T13:44:19.044234459Z I0508 13:44:19.043936       1 sti.go:85] Successfully pushed 172.30.7.46:5000/andy-test/origin-ruby-sample
```

**Comment by liggitt** [comment]
Fixed in #1772
